# Hedef Değişken ve Özellikler Arasındaki İlişkiyi Anlama

### 1. Yorum ve Anlam


**ABD’deki ailelerin yıllık dinlenme/eğlenme harcamalarını bilmek istediğimizi varsayalım. Aşağıdaki modeli tahmin ettik:**



**harcamalar = 873 + 0.0012∗yıllık_gelir + 0.00002∗yıllık_gelir^2 − 223.57∗cocuk_sahibi**




**Harcama ABD doları cinsinden eğlenme yıllık harcamasıdır, yıllık_gelir ABD doları cinsinden yıllık gelirdir ve cocuk_sahibi çocuklu aileleri gösteren değişkendir. **Tahmin edilen katsayıları yorumlayınız. Yorumlarınızın istatistiksel olarak anlamlı olduğundan emin olmak için hangi ek istatistikler verilmelidir? Cevabınızı yazın ve bir Gist'e kaydedin.**

Yıllık gelir miktarı arttıkça eğlence için yıllık harcama miktarı da artmaktadır. 

Ama çocuklu ailelerin eğlence için yaptıkları yıllık harcama azalmaktadır. 

R^2 ile bağımsız değişkenlerin hedef değişkeni açıklamakta yeterli olup olmadığına bakılabilir. %70' in üzerinde ise iyi bir modeldir.

p değerlerine bakarak  bağımsız değişkenlerin modele katkısını anlayabiliriz. 0' a ne kadar yakınsa varyansı o kadar iyi açıklar

### 2. Hava durumu modeli

Bu alıştırmada, önceki dersteki sıcaklık verileriyle çalışacaksınız. Bu ödevi tamamlamak için, gist dosyasına, aşağıdaki görevlere ilişkin çözümlerinizi içeren Jupyter notebook'u bir link olarak ekleyin:

In [11]:
import pandas as pd
import statsmodels.api as sm

In [12]:

weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

**1. Hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu doğrusal bir regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. Tahmin edilen katsayılar istatistiksel olarak anlamlı mıdır? Tahmini katsayılar önceki beklentileriniz doğrultusunda mı? Tahmin edilen katsayıları yorumlayınız. Hedef ve açıklayıcı değişkenler arasındaki ilişkiler nelerdir?**

In [13]:
Y2= weather_df["Apparent Temperature (C)"]-weather_df["Temperature (C)"]
Y2



0       -2.083333
1       -2.127778
2        0.000000
3       -2.344444
4       -1.777778
           ...   
96448    0.000000
96449    0.000000
96450    0.000000
96451    0.000000
96452    0.000000
Length: 96453, dtype: float64

In [14]:
X2 = weather_df[['Humidity','Wind Speed (km/h)']]

In [15]:
import statsmodels.api as sm

X3 = sm.add_constant(X2)
results = sm.OLS(Y2, X3).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 02 May 2021   Prob (F-statistic):               0.00
Time:                        14:38:17   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 2.4381      0.021    115.948      0.000       2.397       2.479
Humidity             -3.0292      0.024   -126.479      0.000      -3.076      -2.982
Wind Speed (km/h)    -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Y2= -2.4381 + 3.0292 * Humidity + 0.1193 * Wind Speed (km/h)**

Nemlilik ve rüzgar hızının p değeri 0.05'ten küçüktür,sıfıra çok yakınıdr. Katsayılar istatistiksel olarak kat sayılar anlamlıdır.

Ancak R^2 çok düşüktür. Yani iyi bir model değildir.

Nemlilik 1 arttığında hedef değişkenimiz 3.0292 derece artmaktadır.

Rüzgar hızı 1 arttığında hedef değişkenimiz 0.1193 derece artmaktadır.

Sadece görünür sıcaklık ile bir model oluşturalım.

In [21]:
Y4 =weather_df['Apparent Temperature (C)']
X4 = weather_df[['Humidity','Wind Speed (km/h)']]

In [23]:
import statsmodels.api as sm

X5 = sm.add_constant(X4)
results = sm.OLS(Y4, X5).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                               OLS Regression Results                               
====================================================================================
Dep. Variable:     Apparent Temperature (C)   R-squared:                       0.402
Model:                                  OLS   Adj. R-squared:                  0.402
Method:                       Least Squares   F-statistic:                 3.242e+04
Date:                      Sat, 01 May 2021   Prob (F-statistic):               0.00
Time:                              14:42:18   Log-Likelihood:            -3.4065e+05
No. Observations:                     96453   AIC:                         6.813e+05
Df Residuals:                         96450   BIC:                         6.813e+05
Df Model:                                 2                                         
Covariance Type:                  nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                40.3055      0.123    328.275      0.000      40.065      40.546
Humidity            -35.4665      0.140   -253.615      0.000     -35.741     -35.192
Wind Speed (km/h)    -0.3132      0.004    -79.219      0.000      -0.321      -0.305
==============================================================================
Omnibus:                     3489.359   Durbin-Watson:                   0.059
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3911.583
Skew:                          -0.469   Prob(JB):                         0.00
Kurtosis:                       3.305   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Y4= 40.3055 - 35.4665 * Humidity - 0.3132 * Wind Speed (km/h)**

sabitlerin p değerleri sıfıra çok yakındır. Katsayılar istatistiksel olarak anlamlıdr.

R^2 bir model için çok düşüktür, iyi bir model değildir.

Nemlilik 1 arttığında görünür sıcaklık 35.4665 derece azalmaktadır.

Rüzgar hızı 1  arttığında görünür nsıcaklık 0.3132 derece azalmaktadır.



**2. Ardından, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Katsayılar istatistiksel olarak anlamlı mıdır? Nem ve rüzgar hızı için tahmini katsayıların işaretleri değişti mi? Tahmin edilen katsayıları yorumlayınız.**

In [35]:
weather_df['Hum_Wind'] = weather_df.Humidity * weather_df["Wind Speed (km/h)"]

In [36]:
X6 = weather_df[['Humidity','Wind Speed (km/h)', 'Hum_Wind']]

In [37]:
import statsmodels.api as sm

X7 = sm.add_constant(X6)
results = sm.OLS(Y2, X7).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sat, 01 May 2021   Prob (F-statistic):               0.00
Time:                        15:04:00   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Hum_Wind              0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Y2= -0.0839 - 0.1775 * Humidity - 0.0905 * Wind Speed (km/h) + 0.2971 * Hum_Wind**

Katsayıların p değerleri 0.05'ten küçük olup sıfıra çok yakındır.

R^2 0.288'den 0.341'e çıktı. bir artış var ama hala model olarak kabul edebilmek için çok kötü.

Eski modelde hem nemliliğin hem de rüzgar  hızının işareti pozitif iken şuan  negatiftir.

Nemlilik 1 arttığında hedef değişkenimiz 0.1775 derece azalmaktadır.

Rüzgar hızı 1 arttığında hedef değişkenimiz 0.0905 derece azalmaktadır.

Nemlilik ve rüzgar hızının birlikte etkisi 1 arttığında hedef değişkenimiz 0.2971 derece artmaktadır.




### 3. Ev fiyatları modeli

Bu alıştırmada, ev fiyatları modelinizi yorumlayacaksınız. Bu ödevi tamamlamak için, gist dosyasına, aşağıdaki görevlere ilişkin çözümlerinizi içeren Jupyter notebook'u bir link olarak ekleyin:

In [2]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

**1. Ev fiyatları modelinizi tekrar çalıştırın ve sonuçları yorumlayın. Hangi özellikler istatistiksel olarak anlamlı ve hangileri değildir?**

In [4]:
Y_hp = house_prices_df['SalePrice']
X_hp= house_prices_df[['OverallQual','YearBuilt', '1stFlrSF', 'GrLivArea']]

In [5]:
import statsmodels.api as sm

X1_hp = sm.add_constant(X_hp)
results = sm.OLS(Y_hp, X1_hp).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.757
Model:                            OLS   Adj. R-squared:                  0.756
Method:                 Least Squares   F-statistic:                     1130.
Date:                Sun, 02 May 2021   Prob (F-statistic):               0.00
Time:                        12:46:10   Log-Likelihood:                -17513.
No. Observations:                1460   AIC:                         3.504e+04
Df Residuals:                    1455   BIC:                         3.506e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -9.683e+05   8.11e+04    -11.946      0.000   -1.13e+06   -8.09e+05
OverallQual  2.381e+04   1135.983     20.960      0.000    2.16e+04     2.6e+04
YearBuilt     449.9097     42.718     10.532      0.000     366.113     533.706
1stFlrSF       35.5350      3.319     10.706      0.000      29.024      42.046
GrLivArea      50.0252      2.729     18.332      0.000      44.672      55.378
==============================================================================
Omnibus:                      402.630   Durbin-Watson:                   1.995
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            33389.524
Skew:                          -0.158   Prob(JB):                         0.00
Kurtosis:                      26.426   Cond. No.                     2.19e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.19e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**SalePrice = -968324.616462354 + 23810.52203049 * OverallQual + 449.90970865 * YearBuilt + 35.53502386 * 1stFlrSF + 50.02524533 * GrLivArea**

R^2 %70 in üstünde, model olarak kabul edilebilir.


Değişkenlerin p değerleri 0  a çok yakınıdr. İstatistiksel olarak anlamlıdır.





**2. Şimdi, anlamsız özellikleri modelinizden hariç tutun. Bir şey değişti mi?**

In [8]:
Y_hp = house_prices_df['SalePrice']
X_hp1= house_prices_df[['OverallQual', '1stFlrSF', 'GrLivArea']]

In [9]:
import statsmodels.api as sm

X2_hp = sm.add_constant(X_hp1)
results = sm.OLS(Y_hp, X2_hp).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.738
Model:                            OLS   Adj. R-squared:                  0.737
Method:                 Least Squares   F-statistic:                     1367.
Date:                Sun, 02 May 2021   Prob (F-statistic):               0.00
Time:                        12:48:50   Log-Likelihood:                -17566.
No. Observations:                1460   AIC:                         3.514e+04
Df Residuals:                    1456   BIC:                         3.516e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const       -1.161e+05   4943.734    -23.487      0.000   -1.26e+05   -1.06e+05
OverallQual  3.046e+04    979.178     31.111      0.000    2.85e+04    3.24e+04
1stFlrSF       39.3711      3.421     11.507      0.000      32.660      46.082
GrLivArea      43.1915      2.749     15.713      0.000      37.799      48.584
==============================================================================
Omnibus:                      378.277   Durbin-Watson:                   1.993
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            23285.492
Skew:                          -0.196   Prob(JB):                         0.00
Kurtosis:                      22.561   Cond. No.                     9.35e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.35e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Y_hp= -1.161e+05 + 3.046e+04 * OverallQual + 39.3711 * 1stFlrSF + 43.1915* GrLivArea**

p değerleri bütün değişkenlerin 0 a yakın olduğu için std err en yüksek olanı çıkardım ama R^2 çok düştü.


Değişkenlerin katsayısında da çok büyük farklılklar vardır.



**3. Ev fiyatları ile olan ilişkilerini belirleyerek istatistiksel olarak anlamlı katsayıları yorumlayın. Ev fiyatları üzerinde hangi özellikler daha belirgin bir etkiye sahiptir?**

Seçtiğim değişkenler bir modelde kullanmak için iyi sayılır, R^2 %70 in üstündedir. 

Bütün değişkenlerin p değerleri 0 a çok yakın ve hepsi istatistiksel olarak anlamlıdır

YearBuilt değişkenini çıkardığımda R^2 çok düştü, ev fiyatında belirleyici bir özelliktir.

**4. Sonuçlar size mantıklı geliyor mu? Değilse, potansiyel nedenleri açıklamaya çalışın.**

p değeri 0.05 'ten küçük değişkeni çıkarınca modelde değişikler meydana geldi. Beklenen bir durumdu.
Bu yüzden sonuçlar mantıklıdır.